etna/lib/etna/client.rb

```diff
@@ -149,46 +183,220 @@ def request_headers
+      end
+
+      def refresh_endpoint
+        "/api/tokens/generate"
+      end
+
+      def request_headers
+        {
+          'Content-Type' => 'application/json',
+          'Accept' => 'application/json, text/*',
+          'Authorization' => "Etna #{@token}"
+        }
+      end
+
+      def retrier
+        @retrier ||= Retrier.new(max_retries: 5, backoff_time: 10, logger: @logger)
       end
     end
 
-    def request(uri, data)
-      if block_given?
-        verify_mode = @ignore_ssl ?
-          OpenSSL::SSL::VERIFY_NONE :
-          OpenSSL::SSL::VERIFY_PEER
-        Net::HTTP.start(uri.host, uri.port, use_ssl: true, verify_mode: verify_mode, read_timeout: 300) do |http|
-          http.request(data) do |response|
+    class Retrier
+      # Ideally the retry code would be centralized with metis_client ...
+      #   unsure what would be the best approach to do that, at this moment.
+      include Etna::ClientUtils
+
+      def initialize(ignore_ssl: false, max_retries: 10, backoff_time: 15, logger:)
+        @max_retries = max_retries
+        @ignore_ssl = ignore_ssl
+        @backoff_time = backoff_time
+        @logger = logger
+      end
+
+      def retry_request(uri, req, retries: 0, &block)
+        retries += 1
+
+        begin
+          @logger.debug("\rWaiting for #{uri.host} restart"+"."*retries+"\x1b[0K")
+
+          sleep @backoff_time
+        end if retries > 1
+
+        if retries < @max_retries
+          begin
+            if block_given?
+              request(uri, req) do |block_response|
+                if net_exceptions.include?(block_response.class)
+                  @logger.debug("Received #{block_response.class.name}, retrying")
+                  retry_request(uri, req, retries: retries, &block)
+                elsif block_response.is_a?(OpenSSL::SSL::SSLError)
+                  @logger.debug("SSL error, retrying")
+                  retry_request(uri, req, retries: retries, &block)
+                else
+                  status_check!(block_response)
+                  yield block_response
+                end
+              end
+            else
+              response = request(uri, req)
+            end
+          rescue OpenSSL::SSL::SSLError => e
+            if e.message =~ /write client hello/
+              @logger.debug("SSL error, retrying")
+              return retry_request(uri, req, retries: retries)
+            end
+            raise e
+          rescue *net_exceptions => e
+            @logger.debug("Received #{e.class.name}, retrying")
+            return retry_request(uri, req, retries: retries)
+          end
+
+          begin
+            retry_codes = ['503', '502', '504', '408']
+            if retry_codes.include?(response.code)
+              @logger.debug("Received response with code #{response.code}, retrying")
+              return retry_request(uri, req, retries: retries)
+            elsif response.code == '500' && response.body.start_with?("Puma caught")
```

Comment:
Hmm, what happens to other 500s here?

Reply:
They should bubble up and get caught by the subsequent call to `status_check!`, which means they would behave the same as before. i.e. 500 "Server Errors" will just feed back to the client, etc. I believe this was put in place for metis_client, because the `Puma caught` is the raw message sent back when Puma is waiting for the app to come back up.